### PR TITLE
feat: validate ice arguments

### PR DESCRIPTION
**Pull request by HaleTom**
# feat: validate ice arguments

## Summary

- Validate ice **values** at parse time (one-time check when `zinit ice` or inline ices are set)
- Invalid values produce a **warning** (not an error) so shells still load
- Catches common typos like `as'none'` instead of `as'null'`
- Works with both quote syntax (`as'null'`) and colon syntax (`as:null`)
- Improves the existing `proto` error message to list valid protocols

## Motivation

Ice **names** are validated — unknown ice names like `foo` produce an error. But ice
**values** are silently accepted regardless of validity. This leads to confusing
failures:

- `as'none'` silently falls through to default sourcing behavior instead of `as'null'`
- `%at-clone` in `atpull` runs as a literal command instead of substituting `atclone`

Fixes #766.

### Validate at ice-set time, not at shell load time

Validation happens in `.zinit-ice()`, which only runs when the user explicitly sets
ices (via `zinit ice ...` or inline with `zinit light`/`zinit load`). It does NOT run
on subsequent shell loads when ices are read from the `._zinit/` disk cache. This
means:

- **One-time cost**: The check runs once when ices are set, not on every shell startup
- **No startup breakage**: Cached ices from older zinit versions continue to work
- **Self-correcting**: Next time the user re-ices a plugin, validation runs

### Warning, not error

Invalid values produce a `{warn}Warning{b-warn}:` message via `+zi-log` but do NOT
abort. The plugin continues to load with the invalid value (which falls through to
default behavior, same as before). This avoids breaking existing configurations.

### Validated ices

| Ice | Valid values | Rationale |
|-----|-------------|-----------|
| `as` | `command`, `program`, `null`, `completion` | Finite, well-defined set. Common typo: `as'none'` instead of `as'null'` |
| `wait` | suffix letter `a`, `b`, `c`, or none | `wait'3d'` silently drops `d`; now warns |
| `depth` | positive integer | `depth'abc'` would fail at git clone time; now warns earlier |

**`proto`**: Already has an error guard in `zinit-install.zsh:423-460` that catches
ALL invalid protocols with `return 1`. Instead of adding a duplicate warning, the
error message was improved to list valid protocols: `"Expected one of: git, http, https, ssh, ftp, ftps, rsync."`

**Not validated:**

- `from` — intentionally accepts arbitrary hostnames as fallback
- `atpull` — complex syntax (`%atclone`, `!` prefix, raw commands) makes reliable validation impractical
- Condition strings in `wait` (e.g., `wait'[[ -n $commands[fzf] ]]'`) — arbitrary shell expressions, not possible to validate at parse time

## Out of Scope

- **Annex ice extensions**: Custom ices from zinit annexes (e.g., `zinit-annex-bin-gem-node`, `zinit-annex-binary-symlink`) are not validated. These are third-party extensions with their own value semantics. Adding validation for annex ices would require each annex to declare its valid values, which is a separate feature.

## Testing

41 tests covering:
- Valid and invalid `as` values (including `as:` colon syntax)
- Valid `wait` numeric formats, condition strings, and invalid suffix letters
- Valid and invalid `depth` values
- Combined ice validation, `--` prefix regression, and `for` syntax integration

### CODE REVIEW

**Comment by HaleTom** [comment]
# Production-readiness review: feat: validate ice arguments

PR: https://github.com/zdharma-continuum/zinit/pull/778
Fixes: #766
Reviewed: 2026-05-20T19:29:58+07:00

## Context files

```
- zinit.zsh:2274-2339 (.zinit-ice + .zinit-validate-ice functions)
- zinit-install.zsh:423-460 (proto validation case block)
- tests/ice-validation.zunit:1-263 (all 41 tests)
- .agents/PR.md:1-69 (PR description and design decisions)
```

## Production-ready bar for this PR

- Implementation matches PR description: validates `as`, `wait`, `depth` at ice-set time
- Warnings (not errors) — shells still load with invalid values
- `as` validation accepts exactly `command`, `program`, `null`, `completion`
- `wait` validation correctly skips condition strings (arbitrary shell expressions)
- `depth` validation rejects non-positive-integer values
- `proto` error message lists all valid protocols
- No early return in `.zinit-ice()` can skip validation
- 41 tests cover valid/invalid values, colon syntax, combined ices, edge cases
- Code is maintainable: clear function boundary, documented intent
- No regressions: existing ice parsing behavior unchanged

## Findings

### 1. Correctness & functional completeness

**NON-BLOCKING** `as'program'` accepted as valid but silently normalised to `command` before validation runs
- Type: Plausible risk
- Evidence: `zinit.zsh:2282` normalises `program` → `command`, then `zinit.zsh:2305` validates
- Confidence: High
- Note: Technically correct but listing `program` as an expected value in the warning message may confuse users who never see it as a final value. Consider either (a) documenting the normalisation in the warning, or (b) not listing `program` since users can't observe it post-normalisation.

**Verified:** `wait` validation correctly skips condition strings — the `<->` pattern fails on `[[ ... ]]` and `(( ... ))` expressions.
- Evidence: `zinit.zsh:2317` — `${w%%[^0-9]*} = <->` requires the value to start with digits

**Verified:** No early return can skip `.zinit-validate-ice()` — function falls through to line 2295 unconditionally.
- Evidence: `zinit.zsh:2274-2296` — `break` at 2279 exits the `for` loop, not the function

### 2. Architecture & boundary integrity

**Verified:** Validation is called at the right point — after ice parsing and normalisation, before return.
- Evidence: `zinit.zsh:2295`

**Verified:** `proto` error message lists exactly the protocols accepted by the case pattern at `zinit-install.zsh:424`.
- Evidence: `zinit-install.zsh:458` — `git, http, https, ssh, ftp, ftps, rsync` matches `(|ftp(|s)|git|http(|s)|rsync|ssh)`

### 3. Code clarity, clean code & maintainability

**NIT:** Function uses `warncreateglobal` option — unnecessary since no globals are created.
- Type: Verified issue (cosmetic)
- Evidence: `zinit.zsh:2302`
- Confidence: Low — this is defensive coding, not harmful

**Verified:** Clear function naming (`.zinit-validate-ice`), documented intent comment, appropriate `localoptions`.

### 4. Comments & code documentation

**Verified:** Function header comment accurately describes behavior: "Validates ice values at parse time. Warns (not errors) about invalid values — behavior is unchanged."
- Evidence: `zinit.zsh:2299-2300`

### 5. Tests & validation

**Verified:** 41 tests confirmed via `grep -c "^@test"`.
- Evidence: `tests/ice-validation.zunit` — 41 `@test` declarations

**NON-BLOCKING:** Missing test for `wait'!3'` (bang prefix, no suffix letter) — should be valid but not explicitly tested.
- Type: Plausible risk
- Confidence: Medium
- Note: The validation logic handles this correctly (strips `!`, sees bare `3`, passes), but an explicit test would prevent regressions.

**Verified:** Tests cover colon syntax (`as:null`, `as:none`), `--` prefix (`--as'null'`), `for` syntax integration, condition strings in `wait`.

### 6. Performance

**Verified:** Validation runs once at ice-set time, not on shell startup. No performance concern.

### 7. Operational risk

**Verified:** Warnings don't abort — plugin loads with invalid value (falls through to default behavior). No breaking change.

### 8. Adversarial review

**Tested:** Could `wait'3.5'` bypass suffix validation? No — `w="${w%%.[0-9]##}"` strips `.5`, leaving `3`, which is valid. Correct.

**Tested:** Could a condition string like `wait'[[ -n $commands[fzf] ]]'` trigger a false positive? No — the condition doesn't match `<->` (not starting with digits). Correct.

**Tested:** Could `depth'0'` pass? Yes — `<->` matches `0`. The PR says "positive integer" but `0` would pass. This is a minor discrepancy.
- Type: Verified issue (minor)
- Evidence: `zinit.zsh:2324` — `<->` matches zero or more digits, including `0`
- Confidence: High
- Note: If `0` should be rejected, change `<->` to `<1->` or add `&& [[ ${ZINIT_ICES[depth]} != 0 ]]`.

## What I could not fully verify

- Runtime behaviour in a live Zsh shell (no Zsh environment available for integration testing)
- Interaction with annex ice extensions (out of scope per PR description)
- Whether `+zi-log` formatting tags render correctly in all terminal emulators

## Final verdict

⚠️ Merge after addressing blocking items.

No blocking issues found. The 3 non-blocking/nit findings and 1 minor verified issue are:

1. `as'program'` listed as valid but normalised before validation — consider documenting or removing from expected values list
2. Missing `wait'!3'` test case — add for regression safety
3. `warncreateglobal` option is unnecessary — cosmetic, harmless
4. `depth'0'` passes validation despite PR claiming "positive integer" — consider `<1->` pattern

All findings are non-blocking. The implementation is correct, well-tested, and safe. The `depth'0'` discrepancy is the most substantive — recommend addressing before merge if `0` is not intended to be valid.

**Comment by HaleTom** [comment]
# Production-readiness review: feat: validate ice arguments (pass 2)

PR: https://github.com/zdharma-continuum/zinit/pull/778
Fixes: #766
Reviewed: 2026-05-20T19:37:53+07:00

## Context files

```
- zinit.zsh:2295-2334 (.zinit-validate-ice — verified all 4 fixes applied)
- zinit-install.zsh:423-460 (proto validation case block — unchanged)
- tests/ice-validation.zunit:1-276 (43 tests, verified new tests present)
- .agents/PR.md:1-69 (PR description)
```

## Production-ready bar for this PR

- Implementation matches PR description: validates `as`, `wait`, `depth` at ice-set time
- Warnings (not errors) — shells still load with invalid values
- `as` validation accepts exactly `command`, `program`, `null`, `completion`
- `wait` validation correctly skips condition strings (arbitrary shell expressions)
- `depth` validation rejects non-positive-integer values (including `0`)
- `proto` error message lists all valid protocols
- No early return in `.zinit-ice()` can skip validation
- 43 tests cover valid/invalid values, colon syntax, combined ices, edge cases
- Code is maintainable: clear function boundary, documented intent
- No regressions: existing ice parsing behavior unchanged

## Findings

All findings from pass 1 resolved:

1. ✅ `warncreateglobal` removed from `.zinit-validate-ice` options — `zinit.zsh:2302`
2. ✅ `program` removed from `as` validation warning expected values — `zinit.zsh:2310`
3. ✅ `depth'0'` now rejected via `<1->` pattern — `zinit.zsh:2324`
4. ✅ `wait'!3'` test case added — `tests/ice-validation.zunit:89`
5. ✅ `depth'0'` test case added — `tests/ice-validation.zunit:195`

No remaining findings.

### 1. Correctness & functional completeness

**Verified:** All `as` values correctly validated. `program` still accepted in the case pattern (normalised at `zinit.zsh:2282`) but no longer listed in the warning message.
- Evidence: `zinit.zsh:2305` and `zinit.zsh:2310`

**Verified:** `wait` validation correctly skips condition strings.
- Evidence: `zinit.zsh:2317`

**Verified:** `depth` now correctly rejects `0` via `<1->` pattern.
- Evidence: `zinit.zsh:2324`

### 2. Architecture & boundary integrity

**Verified:** Validation called at correct point — after normalisation, before return.
- Evidence: `zinit.zsh:2295`

**Verified:** `proto` error message matches case pattern.
- Evidence: `zinit-install.zsh:458`

### 3. Code clarity, clean code & maintainability

**Verified:** `warncreateglobal` removed. Options are now minimal and necessary.
- Evidence: `zinit.zsh:2302`

### 4. Comments & code documentation

**Verified:** Function header comment accurate.
- Evidence: `zinit.zsh:2299-2300`

### 5. Tests & validation

**Verified:** 43 tests confirmed. `wait'!3'` and `depth'0'` tests added.
- Evidence: `tests/ice-validation.zunit:89` and `tests/ice-validation.zunit:195`

### 6. Performance

**Verified:** Validation runs once at ice-set time. No performance concern.

### 7. Operational risk

**Verified:** Warnings don't abort. No breaking change.

### 8. Adversarial review

**Verified:** `depth'0'` no longer passes — `<1->` requires value >= 1. Correct.

**Verified:** `as'program'` still accepted in the case pattern (for backward compat with normalisation) but not advertised in the warning. Correct.

## What I could not fully verify

- Runtime behaviour in a live Zsh shell (no Zsh environment available for integration testing)
- Interaction with annex ice extensions (out of scope per PR description)
- Whether `+zi-log` formatting tags render correctly in all terminal emulators

## Final verdict

✅ Ready to merge — no blocking issues.

All review findings from pass 1 have been addressed. The implementation is correct, well-tested (43 tests), and safe. Validation runs once at ice-set time with warnings (not errors), preserving backward compatibility.

File: zinit.zsh [hunk 1]
```diff
@@ -2292,8 +2292,42 @@ $match[7]}:-${ZINIT[__last-formatter-code]}}}:+}}}//←→}
...
+            +zi-log "{warn}Warning{b-warn}:{rst} {ice}wait{rst} ice received invalid" \
+                "suffix letter {apo}\`{data}${w##[0-9]##}{apo}\`{rst}." \
+                "Expected one of: {data2}a{rst}, {data2}b{rst}, {data2}c{rst}, or none."
+        fi
+    fi
+    if (( $+ZINIT_ICES[depth] )) && [[ -n ${ZINIT_ICES[depth]} ]]; then
+        if [[ ${ZINIT_ICES[depth]} != <-> ]]; then
+            +zi-log "{warn}Warning{b-warn}:{rst} {ice}depth{rst} ice received invalid" \
+                "value {apo}\`{data}${ZINIT_ICES[depth]}{apo}\`{rst}." \
+                "Expected a positive integer."
```
[the remark this answers is not in the document]
Comment (HaleTom) [reply zinit.zsh:2328]: Fixed in 71742a7. Changed `<->` to `<1->` so `depth'0'` is now correctly rejected — matches the "Expected a positive integer" message.

File: zinit.zsh [hunk 2]
```diff
@@ -2292,8 +2292,42 @@ $match[7]}:-${ZINIT[__last-formatter-code]}}}:+}}}//←→}
...
+                    "{data2}program{rst}, {data2}completion{rst}."
+                ;;
+        esac
+    fi
+    if (( $+ZINIT_ICES[wait] )) && [[ -n ${ZINIT_ICES[wait]} ]]; then
+        local w="${ZINIT_ICES[wait]#\!}"
+        w="${w%%.[0-9]##}"
+        if [[ ${w%%[^0-9]*} = <-> && -n ${w##[0-9]##} && ${w##[0-9]##} != [abc] ]]; then
+            +zi-log "{warn}Warning{b-warn}:{rst} {ice}wait{rst} ice received invalid" \
+                "suffix letter {apo}\`{data}${w##[0-9]##}{apo}\`{rst}." \
```
[the remark this answers is not in the document]
Comment (HaleTom) [reply zinit.zsh]: Fixed in 71742a7. Extracted `${w##[0-9]##}` into a local `suffix` variable — used in both the condition check and the log message.

File: tests/ice-validation.zunit [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    assert $state equals 0
+
+}
+
+@test 'wait-invalid-suffix-x' {
+    run zinit ice wait'0x'
+    assert $state equals 0
+    assert $output contains 'wait'
+    assert $output contains 'x'
+    assert $output contains 'a'
```
[the remark this answers is not in the document]
Comment (HaleTom) [reply tests/ice-validation.zunit]: Fixed in 71742a7. Changed assertion to `assert $output contains 'Expected one of: a, b, c'` — matches the actual warning text instead of a single character.

File: tests/ice-validation.zunit [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+@test 'as-colon-invalid' {
+    run zinit ice as:none
+    assert $state equals 0
+    assert $output contains 'as'
+    assert $output contains 'none'
+}
+
+@test 'for-invalid-as' {
+    run zinit for as'none' @zdharma-continuum/null
```
[the remark this answers is not in the document]
Comment (HaleTom) [reply tests/ice-validation.zunit:274]: Fixed in 71742a7. Added `assert $state equals 0` — validation warns but does not error, so exit code should be 0.

File: zinit.zsh (hunk 1 above)
[the remark this answers is not in the document]
Comment (HaleTom) [reply zinit.zsh:2328]: Fixed in 6805ff0. Added `{data2}program{rst}` back to the warning message — the case pattern at line 2305 accepts it for backward compat (normalised at line 2282), so the warning should list it too.

File: tests/ice-validation.zunit (hunk 3 above)
[the remark this answers is not in the document]
Comment (HaleTom) [reply tests/ice-validation.zunit]: Fixed in 6805ff0. Added `assert $output contains 'program'` to the `as-invalid-none` test — verifies the warning lists all accepted values.

File: zinit.zsh [hunk 5]
```diff
@@ -2292,8 +2292,43 @@ $match[7]}:-${ZINIT[__last-formatter-code]}}}:+}}}//←→}
...
+.zinit-validate-ice() {
+    builtin setopt localoptions noksharrays extendedglob typesetsilent noshortloops
+    if (( $+ZINIT_ICES[as] )) && [[ -n ${ZINIT_ICES[as]} ]]; then
+        case ${ZINIT_ICES[as]} in
+            (command|program|null|completion) ;;
+            (*)
+                +zi-log "{warn}Warning{b-warn}:{rst} {ice}as{rst} ice received invalid" \
+                    "value {apo}\`{data}${ZINIT_ICES[as]}{apo}\`{rst}." \
+                    "Expected one of: {data2}null{rst}, {data2}command{rst}," \
+                    "{data2}completion{rst}."
```
[the remark this answers is not in the document]
Comment (HaleTom) [reply zinit.zsh]: Already fixed in 6805ff0 — added `{data2}program{rst}` back to the expected values list in the warning message. See earlier reply in the sibling thread.

**Comment by HaleTom** [comment]
@pschmitt @vladdoster I think this PR is now ready for review when you have the chance -- cheers!

**Comment by pschmitt** [comment]
Thanks 👍
